docker/build-tools/Dockerfile

```diff
@@ -57,30 +62,33 @@ RUN go get github.com/jteeuwen/go-bindata/go-bindata
 RUN go get istio.io/tools/protoc-gen-docs
 RUN go get istio.io/tools/cmd/annotations_prep
 
-# Save some bytes in the container
-RUN upx --lzma /go/bin/*
+# Put the stuff we need in its final output location
+RUN mkdir -p ${OUTDIR}/usr/bin
+RUN mkdir -p ${OUTDIR}/usr/include
+RUN mkdir -p ${OUTDIR}/usr/local/go/bin
+RUN mkdir -p ${OUTDIR}/usr/local/go/pkg/mod
 
-# Cleanup
-WORKDIR /
-RUN apt-get purge -y \
-    autoconf \
-    libc-ares-dev \
-    libssl-dev \
-    libtool \
-    unzip \
-    upx \
-    zlibc \
-    zlib1g-dev \
-    && apt-get autoremove -y \
-    && apt-get clean \
-    && rm -rf /var/lib/apt/lists/* \
-    && rm -fr /home /tmp/* \
-    && go clean -cache \
-    && go clean -modcache
+#RUN install -c ${GOPATH}/bin/* ${OUTDIR}/usr/bin
+RUN cp -aR /tmp/bin/protoc ${OUTDIR}/usr/local/go/bin
+RUN cp -aR /tmp/include/* ${OUTDIR}/usr/include
+RUN cp -aR /go/bin ${OUTDIR}/usr/local/go/bin
+RUN cp -aR /usr/local/go/bin/* ${OUTDIR}/usr/local/go/bin
+
+# Use inplace decompression on the go toolchain binaries
+#RUN upx --lzma ${OUTDIR}/usr/bin/*
+
+FROM golang:1.12 as build_tools_mod
 
 # Setup to run the Go compiler with host-side build cache, test cache, and module cache
+FROM ubuntu:xenial
+COPY --from=build_tools_context /out/ /
+COPY --from=build_tools_mod /usr/local/go/pkg/ /usr/local/go/pkg/
+
+RUN apt-get update
```

Comment:
Combine in a single command with &&

Also, are there some puny bytes to be saved by tacking on:

    && apt-get autoremove -y \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/* \

Reply:
`&&` will not pass hadolint.  Each docker command should be one line - and a context should be used to trim the extra layers.

See followup commit.

Reply:
Interesting as the Docker best practice from their site has:
```
Below is a well-formed RUN instruction that demonstrates all the apt-get recommendations.

RUN apt-get update && apt-get install -y \
    aufs-tools \
    automake \
    build-essential \
    curl \
    dpkg-sig \
    libcap-dev \
    libsqlite3-dev \
    mercurial \
    reprepro \
    ruby1.9.1 \
    ruby1.9.1-dev \
    s3cmd=1.1.* \
 && rm -rf /var/lib/apt/lists/*
```

Reply:
Also the apt-get clean mention by @geeknoid can be ignored on official Ubuntu and Debian images as it is run automatically.

Reply:
I suspect hadolint is just picking on each command should be on the same line, not on one line with multiple &&'s on one line.